Cgroup mount paths

alexis-cousein opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
We have two sites in Europe using Docker in combination with the cgroup hook (sometimes in parallel with PBSPro, i.e. with Docker making containers alongside the PBSPro ones made by the cgroup hook, sometimes cooperatively, with the PBSPro container hook creating cgroups within the per-job ones set up by pbs_cgroups.PY).

Very rarely, we see that the parsing of /proc/mounts makes it pick up different mounts of the cgroup controllers as the mount path to use; most of the time this has few ill effects, but since the mounts do not seem to be permanent and we don't synchronise with the docker daemon messing with them, the cgroup hook picks up a mount path but fails when it wants to actually use it.

The cgroup hook code indeed does not seem to countenance that a controller could be mounted in more than one spot, and certainly not that some of the mount paths encountered could be volatile. Furthermore, it always sets paths up using the last mount encountered for the controller in /proc/mounts, which is a recipe for disaster if we have other software adding mounts and then removing them.

At one site, the mount paths picked up usually refer to paths set up by Docker for individual containers (possibly in preparation for the cgroup controller mounts in the container?) At a second site, we see volatile mounts straight in /sys/fs/cgroup.

If those paths stay valid while an event is being processed it's invisible to the end user. Alas, if the cgroup hook picks up a mount path and then tries to use it when it no longer exists, it fails, with a message like e.g.:

"CgroupConfigError ('Failed to create directory: /sys/fs/cgroup/blkio,cpuacct,memory,freezer/pbspro.slice/pbspro-413268.pbs10.slice/ (ENOMEM)',)"

At that site, inspection of the host after the error was discovered did confirm that /sys/fs/cgroup/blkio,cpuacct,memory,freezer is nowhere to be found, but it must have been there when the execjob_begin event for that job caused the hook to start running.

In the past, for one site we've extended the cgroup config file syntax to allow the site admin to hardcode the cgroup controller mount paths. That seems to have fixed all issues, so there's every chance to believe that the /sys/fs/cgroup/<controller> paths are unharmed and always usable. But of course it's onerous for a site admin to have to specify these. This was not submitted in a pull request, but since the problem is resurfacing at other sites, we probably need to reintroduce this.


#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->
For the new build which I have created to address the issue, when there are competing mounts discovered for a given controller, the hook will always use the shorter one; that will automatically pick up a permanent mount in all use cases I've seen and for all errors that I have seen up to now (note: on some systems like HPE/Cray or HPE/SGI, it may pick up /dev/cpuset or /dev/cgmem to manage cpusets resp. memory cgroup controllers instead of the distribution-standard /sys/fs/cgroup mount, but that is not a problem – these are permanent aliases).

But just in case this would not be caused by extra mounts but by a kernel bug that gives an incorrect path for the existing mount in rare race conditions, I also would like to introduce code to allow the mount paths for the cgroup controllers to be specified explicitly. That way sites could work around even those issues without us having to spin another version of the cgroup hook.

I also defined extra controllers that can be enabled, should Docker play more nicely with us when trying to create child cgroups in controller hierarchies that we do not manage (it may be because we do not create directories for some controllers Docker knows about that it is messing with the cgroup controller mounts). If you enable these controllers, the hook will do nothing except correctly populate the tasks file for them (and destroy them when the job ends).

The selection of the shorter mount and the introduction of "no operation" support for extra controllers is not truly an interface change, but I'll still describe it the design document as well.
#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->
https://openpbs.atlassian.net/wiki/spaces/PD/pages/2750709761/Make+cgroup+hook+robust+in+the+face+of+volatile+cgroup+controller+mounts

#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->
Changes to be tested on TH, results to be added in a note.

There is a new PTL test that mounts the cpuset controller on /dev/tstc and the memory controller on /dev/tstm. It then loads a config file that specifies /sys/fs/cgroup/cpuset to be forced for the cpuset controller.

Expected result: for the memory controller the hook shall use the shorter /dev/tstm path, for the cpuset controller it shall use the path specified in the configuration file.

Note: the code currently includes changes for PRs:
https://github.com/openpbs/openpbs/pull/2370
https://github.com/openpbs/openpbs/pull/2372

to ensure it works on my test system (which has a systemd that does not play nice with master) and that there is a minimal set of spurious test failures.

If those PRs are merged before this one, I'll rebase so that the differences will no longer include changes for these other issues.

<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

alexis-cousein commented:
Test harness tests results are as expected (with failures in *resize_fail tests because of incorrect assumptions the test makes, but the cgroup hook actions seem OK from the MoM logs).

Description: Tests from given sources on platforms CENTOS7, SUSE15, CENTOS8, SLES12, SLES15, UBUNTU1604, UBUNTU1804, UBUNTU2004
Platforms: CENTOS7,SUSE15,CENTOS8,SLES12,SLES15,UBUNTU1604,UBUNTU1804,UBUNTU2004
Profile: cgroups_cousein
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|7273|432|16|0|0|72|344|

alexis-cousein commented:
New PTL test runs as expected -- mounts are added that have shorter mount points; the one that is not overrided in the config file does use the shorter mount path, the other one does use the path specified in the config file (despite the fact it's longer).

2021-04-23 20:03:26,844 INFO     Current config: {
    "cgroup_prefix"         : "pbs_jobs",
    "exclude_hosts"         : [],
    "exclude_vntypes"       : ["no_cgroups"],
    "run_only_on_hosts"     : [],
    "periodic_resc_update"  : true,
    "vnode_per_numa_node"   : false,
    "online_offlined_nodes" : true,
    "use_hyperthreads"      : false,
    "ncpus_are_cores"       : false,
    "manage_rlimit_as"      : true,
    "cgroup" : {
        "cpuacct" : {
            "enabled"            : true,
            "exclude_hosts"      : [],
            "exclude_vntypes"    : []
        },
        "cpuset" : {
            "mount_path"          : "/sys/fs/cgroup/cpuset",
            "enabled"            : true,
            "exclude_cpus"       : [],
            "exclude_hosts"      : [],
            "exclude_vntypes"    : [],
            "mem_fences"         : true,
            "mem_hardwall"       : false,
            "memory_spread_page" : false
        },
        "devices" : {
            "enabled"            : false,
            "exclude_hosts"      : [],
            "exclude_vntypes"    : [],
            "allow"              : [
                "b *:* rwm",
                "c *:* rwm"
            ]
        },
        "hugetlb" : {
            "enabled"            : false,
            "exclude_hosts"      : [],
            "exclude_vntypes"    : [],
            "default"            : "0MB",
            "reserve_percent"    : 0,
            "reserve_amount"     : "0MB"
        },
        "memory" : {
            "enabled"            : true,
            "exclude_hosts"      : [],
            "exclude_vntypes"    : [],
            "soft_limit"         : false,
            "default"            : "100MB",
            "reserve_percent"    : 0,
            "swappiness"         : 0,
            "reserve_amount"     : "2GB",
            "enforce_default"    : true,
            "exclhost_ignore_default" : true
        },
        "memsw" : {
            "enabled"            : false,
            "exclude_hosts"      : [],
            "exclude_vntypes"    : [],
            "default"            : "100MB",
            "reserve_percent"    : 0,
            "reserve_amount"     : "10GB",
            "manage_cgswap"      : true,
            "enforce_default"    : true,
            "exclhost_ignore_default" : true
        }
    }
}
2021-04-23 20:03:33,852 INFO     manager on x7273-34-0: import hook pbs_cgroups {'content-type': 'application/x-config', 'content-encoding': 'default', 'input-file': '/tmp/PtlPbsxouix_ah'}
2021-04-23 20:03:33,852 INFOCLI  x7273-34-0(run_cmd): sudo -H /opt/pbs/bin/qmgr -c import hook pbs_cgroups application/x-config default /tmp/PtlPbsxouix_ah
2021-04-23 20:03:33,882 ERROR    <DshUtils.run_cmd>cmd:/opt/pbs/bin/qmgr -c import hook pbs_cgroups application/x-config default /tmp/PtlPbsxouix_ah err: ["hook 'pbs_cgroups' contents overwritten"]
2021-04-23 20:03:34,453 INFO     mom x7273-34-0 log match: searching for "pbs_cgroups.CF;copy hook-related file request received" - with existence - from 2021/04/23 20:03:31... OK
2021-04-23 20:03:34,454 INFOCLI2 x7273-34-0(run_cmd): sudo -H cat /var/spool/pbs/server_priv/hooks/pbs_cgroups.CF
2021-04-23 20:03:34,470 INFOCLI2 x7273-34-0(run_cmd): sudo -H cat /var/spool/pbs/mom_priv/hooks/pbs_cgroups.CF
2021-04-23 20:03:34,541 INFO     server & mom pbs_cgroups.CF match
2021-04-23 20:03:38,545 INFO     MoM sent signal -HUP
2021-04-23 20:03:38,545 INFOCLI2 x7273-34-0(run_cmd): sudo -H /usr/bin/cat /var/spool/pbs/mom_priv/mom.lock
2021-04-23 20:03:38,561 INFOCLI2 x7273-34-0(run_cmd): sudo -H kill -HUP 5069
2021-04-23 20:03:39,649 INFO     mom x7273-34-0 log match: searching for "hook_perf_stat;label=hook_exechost_startup_pbs_cgroups_.* profile_stop" - using regular expression  - with existence - from 2021/04/23 20:03:36 - No match
2021-04-23 20:03:41,343 INFO     mom x7273-34-0 log match: searching for "hook_perf_stat;label=hook_exechost_startup_pbs_cgroups_.* profile_stop" - using regular expression  - with existence - from 2021/04/23 20:03:36... OK
2021-04-23 20:03:41,347 INFO     MoM sent signal -HUP
2021-04-23 20:03:41,347 INFOCLI2 x7273-34-1(run_cmd): ssh x7273-34-1.x7273-34.th.svc.cluster.local sudo -H /usr/bin/cat /var/spool/pbs/mom_priv/mom.lock
2021-04-23 20:03:42,339 INFOCLI2 x7273-34-1(run_cmd): ssh x7273-34-1.x7273-34.th.svc.cluster.local sudo -H kill -HUP 14796
2021-04-23 20:03:46,035 INFO     mom x7273-34-1 log match: searching for "hook_perf_stat;label=hook_exechost_startup_pbs_cgroups_.* profile_stop" - using regular expression  - with existence - from 2021/04/23 20:03:36... OK
2021-04-23 20:03:46,035 INFO     MoM sent signal -HUP
2021-04-23 20:03:46,035 INFOCLI2 x7273-34-2(run_cmd): ssh x7273-34-2.x7273-34.th.svc.cluster.local sudo -H /usr/bin/cat /var/spool/pbs/mom_priv/mom.lock
2021-04-23 20:03:47,647 INFOCLI2 x7273-34-2(run_cmd): ssh x7273-34-2.x7273-34.th.svc.cluster.local sudo -H kill -HUP 14863
2021-04-23 20:03:51,537 INFO     mom x7273-34-2 log match: searching for "hook_perf_stat;label=hook_exechost_startup_pbs_cgroups_.* profile_stop" - using regular expression  - with existence - from 2021/04/23 20:03:36... OK
2021-04-23 20:03:51,538 INFOCLI2 x7273-34-0(run_cmd): sudo -H mkdir /dev/tstm
2021-04-23 20:03:51,555 INFOCLI2 x7273-34-0(run_cmd): sudo -H mount -t cgroup -o rw,nosuid,nodev,noexec,relatime,seclabel,memory cgroup /dev/tstm
2021-04-23 20:03:51,638 INFOCLI2 x7273-34-0(run_cmd): sudo -H mkdir /dev/tstc
2021-04-23 20:03:51,657 INFOCLI2 x7273-34-0(run_cmd): sudo -H mount -t cgroup -o rw,nosuid,nodev,noexec,relatime,seclabel,cpuset cgroup /dev/tstc
2021-04-23 20:03:55,685 INFO     job: executable set to /opt/pbs/bin/pbs_sleep with arguments: 100
2021-04-23 20:03:55,686 INFOCLI  x7273-34-0(run_cmd): sudo -H -u pbsuser /opt/pbs/bin/qsub -l select=1:ncpus=1:host=x7273-34-0 /tmp/PtlPbsJobScripto690exyw
2021-04-23 20:03:55,711 INFO     submit to x7273-34-0 as pbsuser: job 1.x7273-34-0 OrderedDict([('Resource_List.select', '1:ncpus=1:host=x7273-34-0')])
2021-04-23 20:03:55,711 INFOCLI  job script /tmp/PtlPbsJobScripto690exyw
---
#!/bin/bash
#PBS -joe
sleep 100
---
2021-04-23 20:03:55,711 INFOCLI  x7273-34-0(run_cmd): /opt/pbs/bin/qstat -f 1.x7273-34-0
2021-04-23 20:03:55,740 INFO     expect on server x7273-34-0: job_state = R && substate = 42 job 1.x7273-34-0  got: substate = 41
2021-04-23 20:03:56,743 INFOCLI  x7273-34-0(run_cmd): /opt/pbs/bin/qmgr -c set server scheduling=True
2021-04-23 20:03:56,844 INFOCLI  x7273-34-0(run_cmd): /opt/pbs/bin/qstat -f 1.x7273-34-0
2021-04-23 20:03:56,934 INFO     expect on server x7273-34-0: job_state = R && substate = 42 job 1.x7273-34-0 attempt: 2  got: substate = 41
2021-04-23 20:03:57,526 INFOCLI2 x7273-34-1(run_cmd): ssh x7273-34-1 /usr/bin/cat /proc/meminfo
2021-04-23 20:03:57,935 INFOCLI  x7273-34-0(run_cmd): /opt/pbs/bin/qmgr -c set server scheduling=True
2021-04-23 20:03:57,953 INFOCLI  x7273-34-0(run_cmd): /opt/pbs/bin/qstat -f 1.x7273-34-0
2021-04-23 20:03:57,969 INFO     expect on server x7273-34-0: job_state = R && substate = 42 job 1.x7273-34-0 attempt: 3  got: substate = 41
2021-04-23 20:03:58,970 INFOCLI  x7273-34-0(run_cmd): /opt/pbs/bin/qmgr -c set server scheduling=True
2021-04-23 20:03:58,985 INFOCLI  x7273-34-0(run_cmd): /opt/pbs/bin/qstat -f 1.x7273-34-0
2021-04-23 20:03:58,997 INFO     expect on server x7273-34-0: job_state = R && substate = 42 job 1.x7273-34-0 attempt: 4 ...  OK
2021-04-23 20:03:59,146 INFOCLI2 x7273-34-1(run_cmd): ssh x7273-34-1 df -k /var/spool/pbs
2021-04-23 20:03:59,562 INFO     mom x7273-34-0 log match: searching for "create_job: Creating directory /sys/fs/cgroup/cpuset/pbs_jobs.service/jobid/1.x7273-34-0" - with existence - from 2021/04/23 20:03:53... OK
2021-04-23 20:04:00,132 INFO     mom x7273-34-0 log match: searching for "create_job: Creating directory /dev/tstm/pbs_jobs.service/jobid/1.x7273-34-0" - with existence - from 2021/04/23 20:03:53... OK
2021-04-23 20:04:00,133 INFOCLI2 x7273-34-0(run_cmd): sudo -H umount /dev/tstm
2021-04-23 20:04:00,313 INFOCLI2 x7273-34-0(run_cmd): sudo -H rmdir /dev/tstm
2021-04-23 20:04:00,330 INFOCLI2 x7273-34-0(run_cmd): sudo -H umount /dev/tstc
2021-04-23 20:04:00,364 INFOCLI2 x7273-34-0(run_cmd): sudo -H rmdir /dev/tstc

alexis-cousein commented:
Rebased so that PR #2370 code isn't seen as modified in this PR.

alexis-cousein commented:
Changed the reserve_amount in some CF files used in PTL tests to conform more to the default pbs_cgroups.CF file -- if you test on hosts with less than 2GB of memory (e.g. smaller VMs) otherwise this can lead to spurious failures when the hook tells you that doesn't make sense and refuses to run.

alexis-cousein commented:
Rebased on master -- now the changes only reflect this fix (except for some comments to clarify some earlier changes, plus a change to ensure that tests run on all hosts even when MoM sets an rlimit_as limit when vmem is requested).

alexis-cousein commented:
Unfortunately the rebase made me pick up https://github.com/openpbs/openpbs/pull/2376 , which introduces build errors that the Test Harness does not ignore, and that will continue until https://github.com/openpbs/openpbs/pull/2378 gets merged...

alexis-cousein commented:
Test results on Test Harness:

Description: Tests from given sources on platforms CENTOS7, SUSE15, CENTOS8, SLES12, SLES15, UBUNTU1604, UBUNTU1804, UBUNTU2004
Platforms: CENTOS7,SUSE15,CENTOS8,SLES12,SLES15,UBUNTU1604,UBUNTU1804,UBUNTU2004
Profile: cgroups_cousein
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|7343|432|0|0|0|72|360|

nishiya commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -5129,9 +5199,88 @@ def test_manage_rlimit_as(self):
+                                            'pbs_jobs.service/jobid/%s'
+                                            % jid,
+                                        n='ALL', starttime=begin,
+                                        max_attempts=1)
```
Are we guaranteed that the message will be printed to the log by the time the job is in the R state?  Or should we increase max_attempts to more than 1 for slow test machines?  Same comment for line 5265.

alexis-cousein replied:
Yes, we are guaranteed the message will be printed.

nishiya reviewed (commented):
Just a test question.

nishiya commented:
Also, looks like there were a couple of checks that failed.  Please look into it.  Thanks.

alexis-cousein commented:
Rebased, fixed a few tests with spurious skips, retested:

Description: Rerun All Tests From #7733
Platforms: CENTOS7,SUSE15,CENTOS8,SLES12,SLES15,UBUNTU1604,UBUNTU1804,UBUNTU2004
Profile: cgroups_cousein
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|7738|432|0|0|0|56|376|

bayucan reviewed (approved):
Looks good.

nishiya reviewed (approved):
LGTM  Thanks for your persistence and diligence to make the tests work.

bayucan commented:
Merged #2373 into master.
